pkg/controller/sls/base.go

```diff
@@ -0,0 +1,151 @@
+/*
+
+ Copyright 2021 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package sls
+
+import (
+	"errors"
+
+	sdk "github.com/aliyun/aliyun-log-go-sdk"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/crossplane/provider-alibaba/apis/sls/v1alpha1"
+	slsclient "github.com/crossplane/provider-alibaba/pkg/clients/sls"
+)
+
+// BaseObserve is the common logic for controller Observe reconciling
+func BaseObserve(mg resource.Managed, c slsclient.LogClientInterface) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.SLSProject)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotSLSProject)
+	}
+	projectName := cr.Spec.ForProvider.Name
+	klog.InfoS("observing SLSProject resource", "Name", projectName,
+		"Description", cr.Spec.ForProvider.Description)
+
+	project, err := c.Describe(projectName)
+	if slsclient.IsNotFoundError(err) {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	cr.Status.AtProvider = slsclient.GenerateObservation(project)
+	var upToDate bool
+	if (cr.Spec.ForProvider.Name == project.Name) && (cr.Spec.ForProvider.Description == project.Description) {
+		upToDate = true
+		cr.SetConditions(xpv1.Available())
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: getConnectionDetails(project),
+	}, nil
+}
+
+// BaseCreate is the logic for Create reconciling
+func BaseCreate(mg resource.Managed, c slsclient.LogClientInterface) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.SLSProject)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotSLSProject)
+	}
+	name := cr.Spec.ForProvider.Name
+	description := cr.Spec.ForProvider.Description
+	klog.InfoS("creating SLSProject resource", "Name", name, "Description", description)
+	cr.SetConditions(xpv1.Creating())
```

Comment:
So I need to upgrade to `v0.13.0`?  It doesn't work in version `v0.12.0`.

From the latest release note: 
> #242 Sets Deleting and Creating conditions in the managed reconciler to guard against a consumer forgetting to do so in the respective methods.

It might be better for me to do the upgrade in a separate PR as it will affect managed resource database and providerconfig.

Reply:
Ah, that's right. I see you have raised https://github.com/crossplane/provider-alibaba/pull/64 to address this. Thanks!